File: src/main/java/factory_method/creator/VehicleFactory.java
package factory_method.creator;

import factory_method.product.Vehicle;
import factory_method.product.VehicleType;

public abstract class VehicleFactory {
    public Vehicle orderVehicle(VehicleType vehicleType) {
        final var vehicle = createVehicle(vehicleType);

        vehicle.assembly();
        vehicle.loadShip();

        return vehicle;
    }

    protected abstract Vehicle createVehicle(VehicleType vehicleType);
}

File: kotlin/src/main/kotlin/factory_method/creator/VehicleFactory.kt
package factory_method.creator

import factory_method.product.Vehicle
import factory_method.product.VehicleType

abstract class VehicleFactory {

    fun orderVehicle(vehicleType: VehicleType): Vehicle {
        val vehicle = createVehicle(vehicleType)

        vehicle.assembly()
        vehicle.loadShip()

        return vehicle
    }

    protected abstract fun createVehicle(vehicleType: VehicleType): Vehicle
}